File: ios/Classes/YandexSearch.swift
import CoreLocation
import Flutter
import UIKit
import YandexMapsMobile

public class YandexSearch: NSObject, FlutterPlugin {
  private let pluginRegistrar: FlutterPluginRegistrar!
  private let methodChannel: FlutterMethodChannel!
  private let searchManager: YMKSearchManager!
  private var searchSessions: [Int: YandexSearchSession] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "yandex_mapkit/yandex_search",
      binaryMessenger: registrar.messenger()
    )

    let plugin = YandexSearch(channel: channel, registrar: registrar)

    registrar.addMethodCallDelegate(plugin, channel: channel)
  }

  public required init(channel: FlutterMethodChannel, registrar: FlutterPluginRegistrar) {
    self.pluginRegistrar = registrar
    self.methodChannel = channel
    self.searchManager = YMKSearch.sharedInstance().createSearchManager(with: .combined)

    super.init()

    self.methodChannel.setMethodCallHandler(self.handle)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "searchByText":
      searchByText(call, result)
    case "searchByPoint":
      searchByPoint(call, result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func searchByText(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]
    let sessionId  = params["sessionId"] as! Int
    let session = searchManager.submit(
      withText: params["searchText"] as! String,
      geometry: Utils.geometryFromJson(params["geometry"] as! [String: Any]),
      searchOptions: Utils.searchOptionsFromJson(params["searchOptions"] as! [String: Any]),
      responseHandler: {(searchResponse: YMKSearchResponse?, error: Error?) -> Void in
        self.searchSessions[sessionId]?.handleResponse(searchResponse: searchResponse, error: error, result: result)
      }
    )

    searchSessions[sessionId] = YandexSearchSession(
      id: sessionId,
      session: session,
      registrar: pluginRegistrar,
      onClose: { (id) in self.searchSessions.removeValue(forKey: id) }
    )
  }

  public func searchByPoint(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]
    let sessionId = params["sessionId"] as! Int
    let session = searchManager.submit(
      with: Utils.pointFromJson(params["point"] as! [String: NSNumber]),
      zoom: params["zoom"] as? NSNumber,
      searchOptions: Utils.searchOptionsFromJson(params["searchOptions"] as! [String: Any]),
      responseHandler: {(searchResponse: YMKSearchResponse?, error: Error?) -> Void in
        self.searchSessions[sessionId]?.handleResponse(searchResponse: searchResponse, error: error, result: result)
      }
    )

    searchSessions[sessionId] = YandexSearchSession(
      id: sessionId,
      session: session,
      registrar: pluginRegistrar,
      onClose: { (id) in self.searchSessions.removeValue(forKey: id) }
    )
  }
}

File: android/src/main/java/com/unact/yandexmapkit/YandexSearch.java
package com.unact.yandexmapkit;

import android.content.Context;

import androidx.annotation.NonNull;

import com.yandex.mapkit.search.SearchFactory;
import com.yandex.mapkit.search.SearchManager;
import com.yandex.mapkit.search.SearchManagerType;
import com.yandex.mapkit.search.Session;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class YandexSearch implements MethodCallHandler {
    private final SearchManager searchManager;
    private final BinaryMessenger binaryMessenger;
    @SuppressWarnings({"MismatchedQueryAndUpdateOfCollection"})
    private final Map<Integer, YandexSearchSession> searchSessions = new HashMap<>();

    public YandexSearch(Context context, BinaryMessenger messenger) {
        SearchFactory.initialize(context);

        searchManager = SearchFactory.getInstance().createSearchManager(SearchManagerType.COMBINED);
        binaryMessenger = messenger;
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "searchByText":
                searchByText(call, result);
                break;
            case "searchByPoint":
                searchByPoint(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    @SuppressWarnings({"unchecked", "ConstantConditions"})
    public void searchByText(MethodCall call, MethodChannel.Result result) {
        Map<String, Object> params = ((Map<String, Object>) call.arguments);
        int sessionId = ((Number) params.get("sessionId")).intValue();
        Session session = searchManager.submit(
                (String) params.get("searchText"),
                Utils.geometryFromJson((Map<String, Object>) params.get("geometry")),
                Utils.searchOptionsFromJson((Map<String, Object>) params.get("searchOptions")),
                new YandexSearchListener(result, 0)
        );

        YandexSearchSession searchSession = new YandexSearchSession(
                sessionId,
                session,
                binaryMessenger,
                new SearchCloseListener()
        );

        searchSessions.put(sessionId, searchSession);
    }

    @SuppressWarnings({"unchecked", "ConstantConditions"})
    public void searchByPoint(MethodCall call, MethodChannel.Result result) {
        Map<String, Object> params = ((Map<String, Object>) call.arguments);
        int sessionId = ((Number) params.get("sessionId")).intValue();
        Session session = searchManager.submit(
                Utils.pointFromJson((Map<String, Object>) params.get("point")),
                ((Integer) params.get("zoom")),
                Utils.searchOptionsFromJson((Map<String, Object>) params.get("searchOptions")),
                new YandexSearchListener(result, 0)
        );

        YandexSearchSession searchSession = new YandexSearchSession(
                sessionId,
                session,
                binaryMessenger,
                new SearchCloseListener()
        );

        searchSessions.put(sessionId, searchSession);
    }

    public class SearchCloseListener {
        public void onClose(int id) {
            searchSessions.remove(id);
        }
    }
}
